### PR TITLE
[RW-767][risk=low] Implement a notebook cluster cleanup cron

### PR DESCRIPTION
**Pull request by calbach**
- Update Leo APIs
- Factor out service account access token creation
- Add an endpoint which cleans up running notebook clusters
- Add configuration for the maximum allowed lifetime for a cluster

A follow up change will add the cronjob configuration for this.

### CODE REVIEW

File: api/config/config_local.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "firecloud": {
     "billingAccountId": "014D91-FCB792-33D2C0",
     "billingProjectPrefix": "aou-test-f1-",
+    "clusterMaxAgeDays": 10,
```
Comment (danqrodney) [inline api/config/config_local.json]: Could probably be 5 like test? Any reason to make this 10?

File: api/config/config_stable.json [hunk 2]
```diff
@@ -2,6 +2,7 @@
   "firecloud": {
     "billingAccountId": "014D91-FCB792-33D2C0",
     "billingProjectPrefix": "aou-rw-stable-",
+    "clusterMaxAgeDays": 14,
```
Comment (danqrodney) [inline api/config/config_stable.json]: Is this the value you'd expect to use in prod, too?

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+   * by App Engine cron.
+   *
+   * The cluster deletion policy here aims to strike a balance between enforcing
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. Since a cluster was last accessed, its spent over half its remaining
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: This sorta makes sense. It's a bit more complicated to explain to researchers (and to grok for ourselves) than "your cluster will be deleted after N days once it is idle". Given that we have auto-pause, is it important to take idleness into account here?

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
+    WorkbenchConfig config = configProvider.get();
+    Instant now = clock.instant();
+    Duration maxAge = Duration.ofDays(config.firecloud.clusterMaxAgeDays);
+
+    ClusterApi clusterApi = clusterApiProvider.get();
+    List<Cluster> clusters;
+    try {
+      clusters = clusterApi.listClusters(null, false);
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Is there any pagination or filtering available here? 

Given that we're deleting unused clusters after a while, I wouldn't expect this list to grow big enough for us to worry... just checking.

Might be good to log something here about how many clusters are currently running.

File: api/src/main/resources/notebooks.yaml [hunk 5]
```diff
@@ -854,7 +860,7 @@ definitions:
       masterDiskSize:
         type: integer
         description: |
-          Optional, the size of the disk on the master node. Minimum size is 100GB. If unspecified, default size is 500GB.
+          Optional, the size in megabytes of the disk on the master node. Minimum size is 10GB. If unspecified, default size is 100GB.
```
Comment (danqrodney) [inline api/src/main/resources/notebooks.yaml]: Have we changed to only ask for 10 GB or 25 GB yet? Should we?

File: api/src/main/resources/workbench.yaml [hunk 6]
```diff
@@ -680,6 +680,23 @@ paths:
...
             $ref: "#/definitions/AuditBigQueryResponse"
 
+  /v1/cron/checkClusters:
+    get:
+      security: []
+      tags:
+        - offlineCluster
+        - cron
+      description: >
+        Endpoint meant to be called offline to trigger cluster checks and
```
Comment (danqrodney) [inline api/src/main/resources/workbench.yaml]: When do you plan on running this? Will that be a separate PR?

Seems like we should run it a couple times a day at least, so we don't wind up with clusters that never get upgraded since they are always in use at the time in question.

File: api/src/main/resources/workbench.yaml [hunk 7]
```diff
@@ -680,6 +680,23 @@ paths:
...
+        cleanup. Enforces upgrades for older cluster deployments. May be slow to
+        execute. Only executable via App Engine cronjob.
```
Comment (danqrodney) [inline api/src/main/resources/workbench.yaml]: What makes this only executable via cronjob?

Do we want the option of running this ourselves? (Not sure we need to, just checking.)

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    int unusedDeletes = 0;
+    for (Cluster c : clusters) {
+      String clusterId = c.getGoogleProject() + "/" + c.getClusterName();
+      if (ClusterStatus.UNKNOWN.equals(c.getStatus()) || c.getStatus() == null) {
+        log.warning(String.format("unknown cluster status for cluster '%s'", clusterId));
+        continue;
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: What else will we handle in future?

Is there a status here for "paused" that's different from STOPPED? I think we'd want to delete clusters when they are paused...

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        log.info(String.format(
+            "deleting cluster '%s', unused for %s, more than its remaining lifespan (>%s)",
+            clusterId, formatDuration(unusedDuration), formatDuration(lifeRemaining)));
+        unusedDeletes++;
+      } else {
+        // Don't delete.
+        continue;
+      }
+      try {
+        clusterApi.deleteCluster(c.getGoogleProject(), c.getClusterName());
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: There's a race condition here: we fetch the big list up top, start iterating and deleting clusters, and by the time we get to the deleteCluster call for a cluster, it's actually in use again. So we may want to add a check here before deleting a cluster to make sure it's still not running. (Even better would be to have deleteCluster provide the option to say "don't delete if I'm in use"... but I assume that doesn't exist.)

File: api/config/config_local.json (hunk 1 above)
Comment (blrubenstein) [reply api/config/config_local.json]: I could also picture this as 7 or something, just to keep it on the same scale as stable. (Also, a week seems like a reasonable time?)

Either way, I second Dan that I think it should be about the same time as test. Although I see what you're going for, since development is likely happening locally, and thus we are likely to be reusing clusters. I suspect that five minutes every five days is acceptable cost though.

File: api/config/config_staging.json [hunk 10]
```diff
@@ -2,6 +2,7 @@
   "firecloud": {
     "billingAccountId": "014D91-FCB792-33D2C0",
     "billingProjectPrefix": "aou-rw-staging-",
+    "clusterMaxAgeDays": 5,
```
Comment (blrubenstein) [inline api/config/config_staging.json]: I'd actually say staging can probably be pretty small. Maybe even less than 5 days, just because its being used for QA and should be relatively one-off, correct?

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. Since a cluster was last accessed, its spent over half its remaining
+   *    lifespan idle (as defined by [1]). This dynamic limit means that as the
+   *    upgrade horizon approaches, we'll be more aggressive in upgrading.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: This logic is complicated enough (and the consequences dire enough if we get things wrong) that we probably should add some tests (mocking out ClusterApi).

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 3 above)
Comment (blrubenstein) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Agreed. This seems to be most useful once auto pause is in place for enforcing cluster upgrades, and avoiding zombie clusters. I think it makes sense not to be more aggressive based on idleness.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 8 above)
Comment (blrubenstein) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Stopped is (I believe) the same as paused.

Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: OK. 

It seems to me that the right policy is to feel free to delete clusters in a paused / STOPPED state whenever we think is reasonable, but deleting clusters when they are RUNNING just because they are somewhat old is not very nice... we may just happen to be running our cron job in the middle of your lengthy computation.

Perhaps we should have two day settings: # of days after we delete a paused cluster, and # of days after we delete a running cluster. The latter would be significantly bigger... this should give us several days of runs to attempt to kill the cluster when it was paused, instead of running.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 9 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: This comment was written assuming we would attempt to *not* delete clusters that were running, or at least as aggressively as we'd delete paused clusters... I misread the logic initially. See above.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+import org.pmiops.workbench.notebooks.model.Cluster;
+import org.pmiops.workbench.notebooks.model.ClusterStatus;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+
+/**
+ * Offline cluster API. Handles cronjobs for cleanup and upgrade of older
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: So this says:

https://stackoverflow.com/questions/22859423/combining-google-app-engine-cron-job-and-task-is-starting-google-tasks-as-cron

That there's a ten minute timeout for handling these requests. That should be sufficient unless we get a whole lot of clusters or deletes are super-slow... but either way I'd document that here. 

If you wanted to avoid that timeout you could run this as a background task (but lose the ability to return the info in the response) or I guess we could look into having backend instances for workbench (which we might need at some stage...) Probably not necessary.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 4 above)
Comment (rushtong) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: The leo API doesn't have any paging options that I can see. It does have a way to filter on labels, but that's about it.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 8 above)
Comment (rushtong) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: 👍 to not deleting running clusters.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudConfig.java [hunk 13]
```diff
@@ -28,11 +32,10 @@
   private static final String END_USER_API_CLIENT = "endUserApiClient";
   private static final String ALL_OF_US_API_CLIENT = "allOfUsApiClient";
 
-  private static final String[] BILLING_SCOPES = new String[] {
+  private static final List<String> BILLING_SCOPES = ImmutableList.of(
```
Comment (rushtong) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudConfig.java]: Nice!

File: api/src/main/resources/notebooks.yaml (hunk 5 above)
Comment (calbach) [reply api/src/main/resources/notebooks.yaml]: It was just closed: https://github.com/DataBiosphere/leonardo/issues/388, not sure whether the docs were updated.

File: api/src/main/resources/workbench.yaml (hunk 6 above)
Comment (calbach) [reply api/src/main/resources/workbench.yaml]: Yes - follow-up, per PR description.

On the frequency, I'm not totally sure.

Either (1) run frequently, e.g. every 3h or (2) run this daily in the middle of the night (where people are least likely be to using notebooks)

File: api/src/main/resources/workbench.yaml (hunk 7 above)
Comment (calbach) [reply api/src/main/resources/workbench.yaml]: CronInterceptor picks up the "cron" tag above. We can trigger it manually via the UI. Not sure if there's a programmatic need for this anywhere.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 9 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Added. Ultimately this is just mitigation as we can't do this transactionally (without Leo API support) - this is a matter of reducing the window from 10s of seconds -> 1s of seconds.

File: api/config/config_staging.json (hunk 10 above)
Comment (calbach) [reply api/config/config_staging.json]: I think it should be the same as in test. Feel free to argue for that value being lower/higher.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 12 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Added to comment. I don't think we're going to come close to this, especially if we run this cron frequently.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 8 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Also note: there is no autopausing yet, so I'm simulating this signal by looking at the lastAccessed time; later we can switch this to look at `STOPPED`

File: api/src/main/resources/workbench.yaml (hunk 6 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: I think with the separate settings, running this multiple times a day sounds good.

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            "deleting cluster '%s', exceeded max lifetime @ %s (>%s)",
+            clusterId, formatDuration(age), formatDuration(maxAge)));
+        activeDeletes++;
+      } else if (isIdle && age.toMillis() > idleMaxAge.toMillis()) {
+        log.info(String.format(
+            "deleting cluster '%s', idle with age %s (>%s)",
+            clusterId, formatDuration(age), formatDuration(idleMaxAge)));
+        unusedDeletes++;
+      } else {
+        // Don't delete.
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: It might be useful to spit out how many clusters were idle in general, whether or not we deleted them. (Just as an extra piece of data.)

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
+        continue;
+      }
+
+      Instant created = Instant.parse(c.getCreatedDate());
+      Duration age = Duration.between(created, now);
+      Instant lastUsed = Instant.parse(c.getDateAccessed());
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: When exactly does the dateAccessed field get updated? If I open a notebook and keep the notebook open, does it periodically update this value even if nothing is running? Or is it only if I"m running cells? What if the kernel is doing a long-running operation and I'm not doing anything in the UI?

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 14 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: Done

File: api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java (hunk 15 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java]: My understanding is that it is purely based on the most recent web request coming through the proxy, i.e. any traffic to the Jupyter API from the notebooks UI. It's unclear to me whether a tab open in the background would trigger this, or just typing (but not executing), or if only executing cells would trigger it. I could also imagine this varying between browsers. Asked on the Leo doc for clarification.

Ultimately I believe this value is the same one that Leo is using to trigger pause, and eventually we'll just switch this code to look for the STOPPED status instead of this timestamp. In other words, if this date field doesn't match the behavior we want, then autopause may not match the behavior we want.
